src/typescript-worker/types-fetcher.ts

```diff
@@ -9,244 +9,385 @@ import {ModuleResolver} from './module-resolver.js';
 import {
   parseNpmStyleSpecifier,
-  fileExtension,
   changeFileExtension,
   classifySpecifier,
   resolveUrlPath,
+  trimTrailingSlash,
 } from './util.js';
 
 import type {Result} from '../shared/util.js';
+import type {CachingCdn} from './caching-cdn.js';
+import type {PackageJson, NpmFileLocation} from './util.js';
+import {
+  PackageDependencies,
+  DependencyGraph,
+  NodeModulesDirectory,
+  NodeModulesLayoutMaker,
+} from './node-modules-layout-maker.js';
+
+type PackageName = string;
+type PackageVersion = string;
+type FilePath = string;
+type FileContent = string;
+
+/**
+ * The top-level project. Used as a referrer.
+ */
+const root = Symbol();
 
 /**
  * Fetches typings for TypeScript imports and their transitive dependencies, and
  * for standard libraries.
  */
 export class TypesFetcher {
-  private readonly _moduleResolver: ModuleResolver;
-  private readonly _entrypointTasks: Promise<void>[] = [];
-  private readonly _handledSpecifiers = new Set<string>();
-  private readonly _specifierToFetchResult = new Map<
-    string,
-    Deferred<Result<string, number>>
-  >();
-
-  constructor(moduleResolver: ModuleResolver) {
-    this._moduleResolver = moduleResolver;
-  }
+  private readonly _cdn: CachingCdn;
+  private readonly _importMapResolver: ModuleResolver;
+  private readonly _rootPackageJson: PackageJson | undefined;
 
-  /**
-   * Start fetching type definitions for all bare module specifiers in the given
-   * TypeScript source text. Relative module specifiers are ignored.
-   *
-   * This function returns immediately, but begins an asynchronous walk of the
-   * module import graph to fetch all typings and package.json files that will
-   * be needed in order for TypeScript to type check this file. To access the
-   * results, call {@link getTypings}.
-   */
-  addBareModuleTypings(sourceText: string): void {
-    const fileInfo = ts.preProcessFile(sourceText, undefined, true);
-    for (const {fileName: specifier} of fileInfo.importedFiles) {
-      if (classifySpecifier(specifier) === 'bare') {
-        this._entrypointTasks.push(this._handleBareSpecifier(specifier));
-      }
-    }
-    for (const {fileName: lib} of fileInfo.libReferenceDirectives) {
-      this._entrypointTasks.push(this._addLibTypings(lib));
-    }
-  }
+  private readonly _rootDependencies: PackageDependencies = {};
+  private readonly _dependencyGraph: DependencyGraph = {};
+  private readonly _filesByPackageVersion = new Map<
+    PackageName,
+    Map<PackageVersion, Map<FilePath, Promise<Result<FileContent, number>>>>
+  >();
 
-  /**
-   * Start fetching type definitions for a built-in TypeScript lib, like "dom"
-   * or "esnext".
-   *
-   * This function returns immediately, but begins an asynchronous walk of the
-   * <reference> graph. To access the results, await {@link getTypings}.
-   */
-  addLibTypings(lib: string): void {
-    this._entrypointTasks.push(this._addLibTypings(lib));
+  constructor(
```

Comment:
Non actionable thinking out loud / clarifying. This is a really great class and I especially love the thorough testing.

From reading over the class it seems to be a very stateful class, populated by first construction and then `fetchTypes`. This mutates all the relevant data such that retrieved files can be retrieved.

Are there any invariants that need to be kept in mind when calling the methods that could become subtle mutation errors? For example is it valid to call `fetchTypes` multiple times on the same instance? What if the first one has a 500 status code on one of the requests and on the second call it's a 200 status?

Reply:
Good point --  the current signature implies there's more re-use possible than there actually is, and could be misused. It's actually just a single-use class: one is created for every build, and the `fetchTypes` method is only ever called once.

I've refactored it a little so that the only public API is a static method that ensures `fetchTypes` is only called once. Hopefully this is more clear!

To answer the specific question (which is now impossible, but still interesting): If a request initially was a 500 error, then it would always have the same error code if `fetchTypes` was called multiple times, since we cache the HTTP response status code directly when it is not a 200.

Reply:
Awesome! Also that is interesting!